DataFormats/GeometrySurface/test/gpuFrameTransformTest.cpp

```diff
@@ -51,15 +51,15 @@ int main(void) {
   float ge[6 * size];
 
   auto current_device = cuda::device::current::get();
-  auto d_xl = cuda::memory::device::make_unique<float[]>(current_device, size);
-  auto d_yl = cuda::memory::device::make_unique<float[]>(current_device, size);
+  auto d_xl = cudautils::make_device_unique<float[]>(size, nullptr);
```

Comment:
I'm a bit afraid that the "abuse" would lead to confusion at some point.

I'm thinking (*) to add an overload on the caching allocator that would not take a stream at all (or use the `nullptr` to signify no-stream; although that choice would make it impossible to use the allocator with the default stream), in which case the memory block is truly freed at the destructor of the `unique_ptr` (instead of delaying the "true free" until the work using the memory block has finished). My main challenge is the naming of the smart pointers: using `unique_ptr` for both would likely be confusing (in a sense the current `unique_ptr` could be argued to be confusing as well).

(*) e.g. for caching memory allocations in ESProducts, and to reduce the use of CUDA events in the caching allocator

Reply:
Reading up on the CUDA [documentation](https://docs.nvidia.com/cuda/cuda-runtime-api/stream-sync-behavior.html), there are actually two options for the "default" stream:
  - the "legacy default stream"; this synchronises with all (not non-blocking) streams on the same device
  - the "per-thread default stream"; this is per-thread, and does not synchronise with other streams (except for the legacy one)

Passing `0` or `nullptr` will use either of those behaviours depending on the `nvcc --default-stream` option or the `CUDA_API_PER_THREAD_DEFAULT_STREAM` symbol; the default is the "legacy" stream.

Purely from the API point of view, I would use
  - `cudautils::make_device_unique<T>(size, nullptr);` for the unspecified default stream
  - `cudautils::make_device_unique<T>(size, cudaStreamLegacy);` for the legacy default stream
  - `cudautils::make_device_unique<T>(size, cudaStreamPerThread);` for the per-thread default stream
  - `cudautils::make_device_unique<T>(size);` for the synchronous behaviour

to keep the possibility of passing `nullptr` for the generic default stream.
With that naming scheme, `cudaStreamDefault` makes a lot of sense for the unspecified default stream.

> My main challenge is the naming of the smart pointers: using `unique_ptr` for both would likely be confusing (in a sense the current `unique_ptr` could be argued to be confusing as well).

Then I would suggest `unique_ptr` and `make_device_unique` for the synchronous behaviour, and  something like `async_unique_ptr` and `make_device_async_unique` or `unique_ptr_async` and `make_device_unique_async` for the ones that use a stream ?